ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/AtomicActionRecoveryModule.java

```diff
@@ -234,6 +234,15 @@ private void processTransactionsStatus()
         }
    }
 
+    /**
+     * TODO I don't mind how shouldBlockShutdown() is worked out. Open to better solution :-)
```

Comment:
Is this TODO still valid?

Reply:
If you agree that this is a good logic to implement `shouldBlockShutdown`, I'm happy to remove the TODO :-)

Reply:
I think I would need you to assert it because it's a core part of the feature. But what I can say (having looked a bit at AtomicActionRecoveryModule in your branch today) is I _think_ something else in AtomicActionRecoveryModule might need looking at. IIUC at the end of phase two for this recovery module it's possible the _transactionUidVector could have an element in it that was able to be recovered at https://github.com/jmfinelli/narayana/blob/15c7a3299203777bd9356b7aafa4eb6dd371dbb4/ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/recovery/AtomicActionRecoveryModule.java#L226 and so it's possible this would then need a superfluous cycle to the recovery manager in order that _transactionUidVector would be empty. But does this sound right what I am thinking? I don't think it was mentioned before (I think I thought a full clean scan would be enough).

Reply:
Yep, it does sound right. I thought about that possible extra scan but I didn't do anything to improve it as I thought an extra scan would have been safer. No sure why I thought so though :sweat:

Reply:
I think an extra scan might add 180 seconds to the shutdown but what I am most curious of now is whether we are seeing this behaviour of the additional scan in the test and if not, why not (BTW I think we are not seeing it).

Reply:
Yeah, I think it's happening but we're not "seeing" it. It's happening because another recovery cycle needs to take place before the vector is declared empty, but we're not "seeing" it as we're measuring the commits of the BytemanControlledRecord.

Reply:
Got it. Maybe in the test we should count the calls of the AARM suspend call so we can make sure what we are thinking is happening is what is happening?

Reply:
I created a byteman script to do so. Thanks Tom

Reply:
And I modified AARM to avoid running a useless recovery cycle